### DISCUSSION TITLE
WebSocket becomes unresponsive for slow sensors

### DISCUSSION THREAD

**Discussion by neliel123** [miguelgrinberg/microdot#135] [Q&A]
**IMPORTANT**: If you have a question, or you are not sure if you have found a bug in this package, then you are in the wrong place. Hit back in your web browser, and then open a GitHub Discussion instead. Likewise, if you are unable to provide the information requested below, open a discussion to troubleshoot your issue.

**Describe the bug**
I used WebSocket to turn on some DC motors.  I just noticed that when it takes time for starting or stopping dc motors then the connection gets dropped and MicroDot becomes unresponsive with no error.

**To Reproduce**
```
@app.route('/echo')
@with_websocket
def echo(request, ws):
    while True:
        data = ws.receive()
        // turnOnMotor() # or
        // turnOffMotor()
        ws.send(data)
```

**Expected behavior**
The motor is turned off or On upon receiving the command.  Commands can come in milliseconds differential.

**Additional context**
What is the best way to handle this scenario in MicroDot when the function that you are about to call will take some time and how to debug this problem?

Note: 
If I remove the command that turns on or off the DC motor then there is no issue and WebSocket exchanges are okay.  
If I swap the DC motors with LED then there is no problem as it is a quick thing to do for the microcontroller.

Based on my debugging, the issue is probably that it takes time for motors to turn on or turn off (using PWM) and that delay is impacting the WebSocket message exchange for MicroDot.

I am looking for a possible way to handle this.

Note Again: I have tried all microdot_websocket files. (e.g microdot_websocket.py, microdot_websocket_alt.py microdot_asyncio_websocket.py)

**Comment by miguelgrinberg** [comment]
@neliel123 Is there anything about the "Important" notice at the top of the issue form that is not clear? I would appreciate it if you could respect the rules of this project and not file bugs when you have questions. This time I moved the issue to a discussion for you.

**Reply by neliel123** [reply]
Hey,
Sorry, I actually don't know and have not used discussions in GitHub.  I often ask questions by raising issues.
I have noted this.  I will do this on succeeding questions.
Thanks

**Comment by miguelgrinberg** [comment]
> Based on my debugging, the issue is probably that it takes time for motors to turn on or turn off (using PWM) and that delay is impacting the WebSocket message exchange for MicroDot.

How long does it take for these functions to run? Are you positive that they return? Because if the server becomes unresponsive the most likely explanation is that your functions are hanging.

**Reply by neliel123** [reply]
Hey,
Thanks for the reply.  I powered the motor thru PWM by assigning duty cycle.  These functions run quickly I believe as it is only assigning duty cycles but I am not sure.
```
def turnOnMotor():
    print("turning on motor...")
    motor_pin1.duty_u16(0)
    motor_pin2.duty_u16(65535)
```
The code goes like this.  I removed some parts for clarity but the logic goes like the code above.

When this function is called multiple times thru WebSocket.  I noticed that the WebSocket server becomes unresponsive and the WebSocket connection is dropped from the client.

But if I commented out every motor code then no issue is encountered even if I click some button from my HTML page continuously.
```
def turnOnMotor():
    print("turning on motor...")
    #motor_pin1.duty_u16(0)
    #motor_pin2.duty_u16(65535)
```
The issue is when I am driving motors but if it is a simple LED (on/off) no issue whatsoever.

**Reply by miguelgrinberg** [reply]
You haven't answered my question though. Are you positive that your motor pin calls aren't hanging? That would explain the web server becoming unresponsive. That would mean you have a problem controlling your pins, not a problem with the web server.
